File: java/gadgets/src/main/java/org/apache/shindig/gadgets/MessageBundleSubstituter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.shindig.gadgets;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.URI;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Adds message bundle hangman variable substitution and programmatic JavaScript
 * access to message bundle strings.
 */
public class MessageBundleSubstituter implements GadgetFeatureFactory {
  /**
   * {@inheritDoc}
   */
  public GadgetFeature create() {
    return new MessageBundleSubstituterFeature();
  }
}

class MessageBundleSubstituterFeature extends GadgetFeature {
  private static final MessageBundleParser parser
      = new MessageBundleParser();

  private MessageBundle bundle;

  /**
   * Fetches a message bundle from the spec for the given locale.
   *
   * @param spec Specification of gadget with messages to potentially retrieve
   * @param locale Locale corresponding to the request
   * @return The message bundle, or null if not found
   */
  private GadgetSpec.LocaleSpec getLocaleSpec(GadgetSpec spec,
                                              Locale locale) {
    List<GadgetSpec.LocaleSpec> localeSpecs = spec.getLocaleSpecs();
    for (GadgetSpec.LocaleSpec locSpec : localeSpecs) {
      if (locSpec.getLocale().equals(locale)) {
        return locSpec;
      }
    }
    return null;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void prepare(GadgetView gadget, GadgetContext context,
                      Map<String, String> params) throws GadgetException {
    super.prepare(gadget, context, params);
    Locale locale = context.getLocale();
    GadgetSpec.LocaleSpec localeData = getLocaleSpec(gadget, locale);
    if (localeData == null) {
      // en-all
      localeData = getLocaleSpec(gadget,
                                 new Locale(locale.getLanguage(), "all"));
    }
    if (localeData == null) {
      // all-all
      localeData = getLocaleSpec(gadget, new Locale("all", "all"));
    }

    GadgetServerConfigReader config = context.getServerConfig();

    if (localeData != null) {
      URI uri = localeData.getURI();
      if (uri != null) {
        // We definitely need a bundle, now we need to fetch it.
        bundle = config.getMessageBundleCache().get(uri.toString());
        if (bundle == null) {
          RemoteContent data = null;
          data = config.getContentFetcher().fetch(new RemoteContentRequest(uri),
                                                  context.getOptions());
          if (data.getHttpStatusCode() != RemoteContent.SC_OK) {
            throw new GadgetException(
                GadgetException.Code.FAILED_TO_RETRIEVE_CONTENT,
                String.format("Malformed message bundle URL: %s",
                              uri.toString()));
          }
          bundle = parser.parse(data.getResponseAsString());
          config.getMessageBundleCache().put(uri.toString(), bundle);
        }
      }
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void process(Gadget gadget, GadgetContext context,
                      Map<String, String> params) throws GadgetException {
    super.process(gadget, context, params);
    StringBuilder js = new StringBuilder();
    int moduleId = gadget.getId().getModuleId();
    Locale locale = context.getLocale();

    String setLangFmt = "gadgets.prefs_.setLanguage(%d, \"%s\");";
    String setCountryFmt = "gadgets.prefs_.setCountry(%d, \"%s\");";

    js.append(String.format(setLangFmt, moduleId, locale.getLanguage()));
    js.append(String.format(setCountryFmt, moduleId, locale.getCountry()));

    if (null != bundle) {
      gadget.setCurrentMessageBundle(bundle);
      gadget.getSubstitutions().addSubstitutions(Substitutions.Type.MESSAGE,
                                                 bundle.getMessages());

      RenderingContext rc = context.getRenderingContext();
      if (rc == RenderingContext.GADGET) {
        Map<String, String> msgs = bundle.getMessages();
        JSONObject json = new JSONObject();
        try {
          for (Map.Entry<String, String> entry : msgs.entrySet()) {
            json.put(entry.getKey(), entry.getValue());
          }
        } catch (JSONException e) {
          throw new RuntimeException(e);
        }

        String setMsgFmt = "gadgets.prefs_.setMsg(%d, %s);";
        js.append(String.format(setMsgFmt, moduleId, json.toString()));
      }
    }
    gadget.addJsLibrary(JsLibrary.create(JsLibrary.Type.INLINE, js.toString()));
  }
}

File: php/gadgets/src/MessageBundleSubstituter.php
<?
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 * 
 */

class MessageBundleSubstituter extends GadgetFeatureFactory {
	public function create()
	{
		return new MessageBundleSubstituterFeature();
	}
}

class MessageBundleSubstituterFeature extends GadgetFeature {
	private $parser;
	private $bundle = null;
	
	public function __construct()
	{
		$this->parser = new MessageBundleParser();
	}
	
	private function getLocaleSpec($spec, $locale)
	{
		$localeSpecs = $spec->getLocaleSpecs();
		foreach ( $localeSpecs as $locSpec ) {
			//fix me
			if ($locSpec->getLocale()->equals($locale)) {
				return $locSpec;
			}
		}
		return null;
	}
	
	public function prepare($gadget, $context, $params)
	{
		$locale = $context->getLocale();
		// en-US
		$localeData = $this->getLocaleSpec($gadget, $locale);
		if ($localeData == null) {
			// en-all
			$localeData = $this->getLocaleSpec($gadget, new Locale($locale->getLanguage(), "all"));
		}
		if ($localeData == null) {
			// all-all
			$localeData = $this->getLocaleSpec($gadget, new Locale("all", "all"));
		}
		if ($localeData != null) {
			$uri = $localeData->getURI();
			if ($uri != null) {
				// We definitely need a bundle, now we need to fetch it.
				// Doing it a little different then the java version since our fetcher and cache are intergrated
				$fetcher = $context->getHttpFetcher();
				$response = $fetcher->fetch(new remoteContentRequest($uri));
				//TODO caching the parsed bundle instead of just the xml data would be a lot more efficient and speedy
				$this->bundle = $this->parser->parse($response->getResponseContent());				
			}
		}
	}
	
	public function process($gadget, $context, $params)
	{
		$js = '';
		$moduleId = $gadget->getId()->getModuleId();
		$locale = $context->getLocale();
		$setLangFmt = "gadgets.prefs_.setLanguage(%d, \"%s\");";
		$setCountryFmt = "gadgets.prefs_.setCountry(%d, \"%s\");";
		$js .= sprintf($setLangFmt, $moduleId, $locale->getLanguage());
		$js .= sprintf($setCountryFmt, $moduleId, $locale->getCountry());
		if ($this->bundle != null) {
			$gadget->setCurrentMessageBundle($this->bundle);
			$gadget->getSubstitutions()->addSubstitutions('MSG', $this->bundle->getMessages());
			$rc = $context->getRenderingContext();
			if ($rc == 'GADGET') {
				$msgs = $this->bundle->getMessages();
				$json = array();
				foreach ( $msgs as $key => $value ) {
					$json[$key] = $value;
				}
				$setMsgFmt = "gadgets.prefs_.setMsg(%d, %s);";
				$js .= sprintf($setMsgFmt, $moduleId, json_encode($json));
			}
		}
		$gadget->addJsLibrary(JsLibrary::create('INLINE', $js));
	}
}